enhancements/cluster-logging/cluster-logging-log-forwarding.md

```diff
@@ -0,0 +1,309 @@
+last-updated: 2019-09-17
+status: implementable
+see-also:[]
+replaces:[]
+superseded-by:[]
+---
+
+# cluster-logging-log-forwarding
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+The purpose of log forwarding is to provide a declarative way by which adopters of cluster logging can ship
+container and node logs to destinations that are not necessarily managed by the OKD cluster logging infrastructure. This is on-cluster endpoints like the Elasticsearch cluster deployed by cluster logging, an organization's Kafka message bus, or off-cluster to existing log aggregation solutions (e.g. Splunk, Graylog).  Additionally, log forwarding enables administrators to deploy only the components needed to ship the logs without needing to deploy the entire cluster logging stack.
+
+## Motivation
+
+Organizations desire to reuse their existing enterprise log solutions to store container logs.  Providing a declarative mechanism by which administrators define a log destination simplifies their operational burden.  The cluster logging stack is able to deploy a collector to each OKD node that is configured with the necessary permissions to collect the logs, add container metadata (e.g. labels) and ship them to the specified endpoint.
+
+
+
+### Goals
+The specific goals of this proposal are:
+
+* Provide an API that identifies log sources and their desired destination
+* Send logs to an Elasticsearch cluster not managed by the cluster logging infrastructure
+* Send logs to a fluentd instance not managed by the cluster logging infrastructure via fluentd's secureforward protocol
+* Allow deployment of log forwarding without deploying the entirety of the cluster logging infrastructure (e.g. Kibana, Elasticsearch)
+* Support TLS between the collector and destination if so configured
+* Specifically only forward "application" and "infrastructure" logs
+* Re-use log forwarding to write logs to a cluster logging managed log aggregator (e.g. Elasticsearch)
+
+We will be successful when:
+
+* an administrator is able to deploy their own log aggregation service
+* specifies this service in log forwarding configuration
+* the service receives logs from a OKD cluster logging deployed collection agent  
+
+
+
+### Non-Goals
+
+Log forward intends to be opinionated about the source of log messages; it will explicitly define what constitutes "application logs" or "infrastructure logs". It will not allow configuration of additional sources without further design considerations and how those sources may impact the collecting agent. Log forwarding is not intended to provide a complex routing solution as one might achieve by using a custom collector configuration or a messaging solution (e.g. kafka).
```

Comment:
more an implementation detail, but couldn't we accomplish this if we have a comma delimited list and some constants as `source` choices? 

Something like:
```
- source: APPS, INFRA
  destination: abc
- source: APPS, !noisy-namespace
  destination: xyz
- source: INFRA_NODE, INFRA_CONTAINER
  destination: jkl
```

Reply:
opinionated is ok for the first pass, but i think we should be heading towards a config api that allows the admin to specify specific namespaces or even pods (by label for example) that should be collected and where they should be sent.

And this ties back to my earlier question about whether forwarding to the "managed" elastic storage will be configured via this same mechanism, or it will have unique/hardcoded behavior.

Reply:
Part of being opinionated here is that we will not support _every_ configuration option provided by the plugin.  For example https://docs.fluentd.org/output/forward has many parameters - I don't think we want to expose all of them in the CR - for another example, you can specify some of them multiple times - fluentd will load balance/round robin among a list of hosts you specify, each with their own possibly different port,username,password,shared_key, tls cert/key/settings, etc.

If we go down that road, we might as well let the user provide us with the actual fluentd plugin config in that format, as a configmap, and one or more secrets, then the user would provide the configmap and secret name(s) in the clusterlogging CR.

Reply:
> more an implementation detail, but couldn't we accomplish this if we have a comma delimited list and some constants as `source` choices?
> 
> Something like:
> 
> ```
> - source: APPS, INFRA
>   destination: abc
> - source: APPS, !noisy-namespace
>   destination: xyz
> - source: INFRA_NODE, INFRA_CONTAINER
>   destination: jkl
> ```

We could but this is really just a variant of what is proposed: its `[]sources->dest` instead of `source->[]dest`.  I prefer sources being an explicit API choice that define a very specific thing we are collecting instead of a more wide open list.  I think it is likely easier to implement as is now because of how we generate the config; there is no need to figure how to merge sources

Reply:
> opinionated is ok for the first pass, but i think we should be heading towards a config api that allows the admin to specify specific namespaces or even pods (by label for example) that should be collected and where they should be sent.
> 
I think we could accomplish this in future by adding sub-elements to the sources which may define: throttling, exclusions.  Maybe because sources are able to very we should consider defining the sources with a 'type' so they can also be independently configured.

> And this ties back to my earlier question about whether forwarding to the "managed" elastic storage will be configured via this same mechanism, or it will have unique/hardcoded behavior.

Yes. Introduce logic to normalize the `clusterlogging` resource and treat the managed instance as any other.  The initial PR does exactly this.

Reply:
> We could but this is really just a variant of what is proposed: its `[]sources->dest` instead of `source->[]dest`. I prefer sources being an explicit API choice that define a very specific thing we are collecting instead of a more wide open list. I think it is likely easier to implement as is now because of how we generate the config; there is no need to figure how to merge sources

I meant more that we could have constants for different sources, similar to what you had proposed as well - I think you had `logs.app` and `logs.infra` but I'm also proposing that we can append a `!noisy-namespace` to remove it from `logs.app` or something. This way we could cover the config api that Ben referred to above.

Reply:
An alternate API that we might only partially implement in this pass:
```
apiVersion: "logging.openshift.io/v1alpha1"
kind: "ClusterLogging"
spec:
  collection:
    type: fluentd
    forwarding:
      targets:
       - type: “elasticsearch”
         name: elasticsearch
         endpoint: https://elasticsearch.svc.messaging.cluster.local:9200
         certificates:
            secretName: elasticsearch
       - type: “secureforward”
         name: secureforward-offcluster
         endpoint: secureforward.offcluster.com:8586
         certificates:
            secretName: secureforward
      sources:
        logs.app:
          excluded:
          - myproject
          - yourproject
          targets:
          - elasticsearch
          - secureforward-offcluster
        logs.infra:
          throttle:
          - namespace: openshift-registry
            value:  50
          targets:
          - elasticsearch
```